sycl/test/include_deps/sycl_accessor.hpp.cpp

```diff
@@ -76,6 +76,7 @@
 // CHECK-NEXT: range.hpp
 // CHECK-NEXT: info/info_desc.hpp
 // CHECK-NEXT: ext/oneapi/experimental/device_architecture.hpp
+// CHECK-NEXT: ext/oneapi/experimental/forward_progress.hpp
```

Comment:
Why is necessary?

Reply:
info/info_desc.hpp needs info about forward progress enums so it includes forward_progress.hpp which means it will appear as a transitive dependecy.

Reply:
Are you saying that the entire extensions is necessary as part of "core" SYCL include? I can't agree with that, because then our compile time will be skyrocketing going forward with any new PR and it's already terrible.

Reply:
I don't know whether it necessary as part of "core" SYCL but I do know that it is necessary in the files where I have included it, at least to the best of my knowledge. During pre-commit testing, some tests were failing because they were expecting to see my header file as a (direct or indirect) dependency for certain headers. So I added these check lines to make these tests pass. Perhaps this points to a larger systemic issue with the way we've designed our tests/code but that will need to be handled separately in my opinion.

Reply:
>  but that will need to be handled separately in my opinion.

How do you suggest we find a person who would be refactoring the entire SYCL implementation if all other people would be contributing to the constantly growing compile time using that excuse?

Reply:
Just noticed this conversation when I saw f783279655030d27ae94e9a510152c96c221d94b.  The forward progress extension does not document that the application must include a special header file to use it, therefore, the assumption is that you only need to include `<sycl/sycl.hpp>`.  This is the same for all of our other extensions.  I don't think this extension should be different from all the others.

I agree that we should have a more granular set of headers in SYCL.  I think the right way to do this is to keep `<sycl/sycl.hpp>` as the "uber" header that includes everything, and then we can also create a set of more granular headers that include specific features.  We can also create separate headers for each extension (to the extent possible) as part of this effort.

Reply:
Thanks for the clarification @gmlueck !

Reply:
@aelovikov-intel 

> Just noticed this conversation when I saw [f783279](https://github.com/intel/llvm/commit/f783279655030d27ae94e9a510152c96c221d94b). The forward progress extension does not document that the application must include a special header file to use it, therefore, the assumption is that you only need to include `<sycl/sycl.hpp>`. This is the same for all of our other extensions. I don't think this extension should be different from all the others.
> 
> I agree that we should have a more granular set of headers in SYCL. I think the right way to do this is to keep `<sycl/sycl.hpp>` as the "uber" header that includes everything, and then we can also create a set of more granular headers that include specific features. We can also create separate headers for each extension (to the extent possible) as part of this effort.

The second paragraph of the quoted reply summarizes where I was going at when I said that we might have a more systemic issue. For now, however, I'm just respecting the status quo as previous extensions have done. In each of the files that forward_progress.hpp is included, the inclusion is necessary and compilation fails otherwise.

Reply:
> I agree that we should have a more granular set of headers in SYCL. I think the right way to do this is to keep `<sycl/sycl.hpp>` as the "uber" header that includes everything, and then we can also create a set of more granular headers that include specific features. We can also create separate headers for each extension (to the extent possible) as part of this effort.

And I wasn't talking about `sycl/sycl.hpp`, it was about `sycl/detail/core.hpp` (see https://github.com/intel/llvm/pull/12890 that introduced it and there was a bunch of changes after). The problem we have is that it's way too easy to add all the new stuff in such a way that `sycl::handler`/`sycl::queue` depend on it unconditionally and then there is no way to make *any* meaningful separation of features.

Reply:
The relevant include chain is core.hpp includes queue.hpp includes info_desc_helpers.hpp includes info_desc.hpp includes forward_progress.hpp.
I'll see tomorrow if I can perhaps skip the inclusion of forward_progress.hpp from info_desc.hpp without deviating too much from our usual code structure.

Reply:
It's not as much about that header but more about all the templated code that FE has to parse that you put into other headers.

Reply:
I see what you mean. As a worst case scenario, from a client's perspective, a user that just wants to perform basic matrix operations using SYCL does not need anything related to this extension and yet by including sycl/sycl.hpp they will incur some of the compile-time cost of the template parsing that happens even when the templates are not instantiated. 

As @gmlueck said above, it would be ideal to separate headers for big extensions but that could be hard to do. The bigger an extension, the more coupled it will be with existing code. For example, due to a stipulation in the spec of this extension about verifying the properties required by a kernel, I had to go and augment the implementation of the already existing function processProperties in handler.hpp. The alternative, in such cases, would be to add preprocessor conditionals enabling/disabling compilation of extension code.

Reply:
> For example, due to a stipulation in the spec of this extension about verifying the properties required by a kernel, I had to go and augment the implementation of the already existing function processProperties in handler.hpp.

Do you mean this part of the spec:

> If a kernel is submitted to a device that cannot satisfy the request for specific progress guarantees, the implementation must throw an `exception` with the `errc::feature_not_supported` error code.

Properties are getting integrated into a lot of APIs, so I think many headers will end up needing to include the base definitions for properties.

> The alternative, in such cases, would be to add preprocessor conditionals enabling/disabling compilation of extension code but this can quickly become unwieldy.

I'm not sure this is a good path.  We previously considered adding a command line switch that would enable/disable each property, but we rejected this because it introduced too many testing combinations.

Reply:
I agree. I do think that its better to continue doing what we've been doing and perhaps slowly start to separate headers for extensions where it is possible to do so.

Reply:
Tagging @AlexeySachkov for input on this discussion.

Reply:
@aelovikov-intel Do we have your approval on this?
Otherwise, we can put this on hold while an outline of a standard approach on how to deal with these concerns is documented somewhere. I don't want to resort to ad-hoc approaches.

Reply:
Please correct me if I am wrong, @aelovikov-intel, but your point here is not that including into `core.hpp` affects users, but rather that we keep `core.hpp` as bare-bones as possible to help reduce compilation overhead in our tests, as the compiler has to re-parse our headers for even small tests and the less overhead we can have the better.

From a user's perspective, all they know exists from the SYCL spec is `<sycl/sycl.hpp>` and we make no exception for extensions currently. I would personally love to see that change in the future, but for now `sycl.hpp` must include everything and `core.hpp` should include only the essentials.

Reply:
> Please correct me if I am wrong, @aelovikov-intel, but your point here is not that including into core.hpp affects users, but rather that we keep core.hpp as bare-bones as possible to help reduce compilation overhead in our tests

Kind of, see below.

> Perhaps this points to a larger systemic issue with the way we've designed our tests/code but that will need to be handled separately in my opinion.

Yes, there is one, but we *do* work on it. https://github.com/intel/llvm/pull/12890 introduced `<sycl/detail/core.hpp>` and later https://github.com/intel/llvm/pull/13210 introduced an in-tree test supposed to keep the growth of the newly introduced header file under control. As such, saying 

> For now, however, I'm just respecting the status quo as previous extensions have done.

and implying that nothing is getting fixed outside of this PR is incorrect. Note, that if would be impossible for 1-2 people to try to gradually improve things not working full time on that task if other dozens of contributors would just keep bloating `handler.hpp`/`queue.hpp` (that are part of `detail/core.hpp` for obvious reasons) as being done in this PR.

Now, switching to what can be done here. I think you can use some kind of visitor pattern. `<oneapi_properties>::has_property` has to have a loop inside its implementation. What you can do is this:

```
handler.hpp:
// SFINAE-visitor
template <Property>
struct ProcessForwardProgressProperty {
  void operator() {
    // do nothing for non-forward-progress property
  }
};
...
 loop over all the properties passed to the method and call the above `operator()` with it.

forward_progress_property:
// define property and provide partial specialization for the above helper.
```

Reply:
@aelovikov-intel 
I discussed this with @AlexeySachkov and we decided to open an issue regarding this: https://github.com/intel/llvm/issues/13774. I've assigned myself to it and will start working on it immediately.
In the meantime, let's proceed with getting this PR merged.